Mage/Archivist/Archivist.cs

```diff
@@ -54,8 +76,166 @@ public class Archivist : EventEmitter<JObject> {
 	}
 
 
+	// Returns cache value if it exists and has not passed max allowed age
+	private VaultValue GetCacheValue(string cacheKeyName, int? maxAge = null) {
+		if (!_cache.ContainsKey(cacheKeyName)) {
+			return null;
+		}
+
+		VaultValue value = _cache[cacheKeyName];
+		double timespan = (DateTime.Now - value.writtenAt).TotalMilliseconds;
+		if (maxAge != null && timespan > maxAge * 1000) {
+			return null;
+		}
+
+		return value;
+	}
+	
+
+	// Ensures a cache value exists then sets it based on info provided
+	private void SetCacheValue(JObject info) {
+		string topic = (string)info["key"]["topic"];
+		JObject index = (JObject)info["key"]["index"];
+		string cacheKeyName = GetCacheKey(topic, index);
+		VaultValue cacheValue = GetCacheValue(cacheKeyName);
+		if (cacheValue == null) {
+			cacheValue = new VaultValue(topic, index);
+			_cache.Add(cacheKeyName, cacheValue);
+		}
+
+		JObject rawValue = (JObject)info["value"];
+		if (rawValue != null) {
+			ValueSet(topic, index, rawValue["data"], (string)rawValue["mediaType"], (int?)info["expirationTime"]);
+		} else {
+			ValueDel(topic, index);
+		}
+	}
+
+
+	// Return cache dictionary
+	public Dictionary<string, VaultValue> GetCache() {
+		return _cache;
+	}
+
+
+	// Clear out the cache entirely
+	public void ClearCache() {
+		_cache = new Dictionary<string, VaultValue>();
```

Comment:
i wonder if clear is thread safe
the issue here is that clear could be deleting record by record which can break in threading
however if we create a new object and change the reference, we're guaranteed a new object